### ISSUE TITLE
C417 auto-fixer breaks code with late-binding

### ISSUE THREAD

**Issue by mrcljx** [astral-sh/ruff#5502]
Version: ruff 0.0.276 (no regression though)

```diff
  nums = range(4)

- callbacks = map(lambda x: lambda: x, nums)
+ callbacks = (lambda: x for x in nums)
```

Now `list(callbacks)[0]()` will change it's result:

```diff
- 0
+ 3
```

Additionally, a new error appears (pointing out the issue):

> B023 Function definition does not bind loop variable `x`

My expectation is that the either the rule doesn't trigger in this case, or at least the auto-fixer doesn't break code.

**Comment by charliermarsh** [comment]
Is this the "real" code that triggered the violation? It'd be helpful to have some more examples to inform the fix.

**Comment by mrcljx** [comment]
A real (but harder to grok) code example is this:

```py
infos = map(
    lambda aspect: ColumnInfo(
        aspect.value,
        ma.fields.Enum(models.Resolution, required=True, by_value=True),
        AttributeImpl(
            lambda d, ctx: d.aspect_resolutions(aspect).effective(
                ctx.aspects.device_resolution_quota(aspect).fallback
            ),
            display_name="",
        ),
    ),
    models.Aspect,
)
```

which got transformed to

```py
infos = (
    ColumnInfo(
        aspect.value,
        ma.fields.Enum(models.Resolution, required=True, by_value=True),
        AttributeImpl(
            lambda d, ctx: d.aspect_resolutions(aspect).effective(
                ctx.aspects.device_resolution_quota(aspect).fallback
            ),
            display_name="",
        ),
    )
    for aspect in models.Aspect
)
```

for which then the following problem was caught:

> B023 Function definition does not bind loop variable `aspect`

---

The problem effectively happens if the iterator variable is used inside a `lambda`, so a some logic like the following would would be needed:

```
for lambda_node in find_lambda_nodes(map_node):
  if loop_variable_name in find_all_names(lambda_node):
    # loop variable is late-bound to a lambda, using comprehension would introduce bugs, bail
    return
```

**Comment by charliermarsh** [comment]
Thanks, yeah, this makes sense.
